californium-core/src/main/java/org/eclipse/californium/core/network/stack/BlockwiseLayer.java

```diff
@@ -139,39 +159,38 @@ public BlockwiseLayer(final NetworkConfig config) {
 	@Override
 	public void sendRequest(final Exchange exchange, final Request request) {
 
-		BlockOption block2 = request.getOptions().getBlock2();
-		if (block2 != null && block2.getNum() > 0) {
-			// This is the case if the user has explicitly added a block option
-			// for random access.
-			// Note: We do not regard it as random access when the block num is
-			// 0. This is because the user might just want to do early block
-			// size negotiation but actually wants to receive all blocks.
-			LOGGER.fine("request contains block2 option, creating random-access blockwise status");
-			BlockwiseStatus status = new BlockwiseStatus(getSizeForSzx(block2.getSzx()), request.getOptions().getContentFormat());
-			status.setCurrentSzx(block2.getSzx());
-			status.setCurrentNum(block2.getNum());
-			status.setRandomAccess(true);
-			exchange.setResponseBlockStatus(status);
-			lower().sendRequest(exchange, request);
-
-		} else if (requiresBlockwise(request)) {
-			// This must be a large POST or PUT request
-			startBlockwiseUpload(exchange, request);
+		Request requestToSend = request;
 
-		} else {
-			// no blockwise transfer required
-			exchange.setCurrentRequest(request);
-			lower().sendRequest(exchange, request);
+		if (isTransparentBlockwiseHandlingEnabled()) {
+
+			BlockOption block2 = request.getOptions().getBlock2();
+			if (block2 != null && block2.getNum() > 0) {
+				// This is the case if the user has explicitly added a block option
+				// for random access.
+				// Note: We do not regard it as random access when the block number is 0.
+				// This is because the user might just want to do early block
+				// size negotiation but actually want to retrieve the whole body by means of
+				// a transparent blockwise transfer.
+				LOGGER.fine("outbound request contains block2 option, creating random-access blockwise status");
+				addRandomAccessBlock2Status(exchange, request, block2);
+
+			} else if (requiresBlockwise(request)) {
+				// This must be a large POST or PUT request
+				requestToSend = startBlockwiseUpload(exchange, request);
+
+			}
 		}
+
+		exchange.setCurrentRequest(requestToSend);
+		lower().sendRequest(exchange, requestToSend);
 	}
 
-	private void startBlockwiseUpload(final Exchange exchange, final Request request) {
+	private Request startBlockwiseUpload(final Exchange exchange, final Request request) {
 
-		BlockwiseStatus status = findRequestBlockStatus(exchange, request);
+		final KeyUri key = getKey(exchange, request);
+		Block1BlockwiseStatus status = getOutboundBlock1Status(key, exchange, request);
```

Comment:
(Just to be sure,  we talk about client and server CoAP)

> Well, you would accept a response from an intermediary proxy instead of a server as well, wouldn't you?

I don't see the problem here, I received a response from the server I connected to. (proxy or not)

> So far, canceling was something the client code had to do deliberately, wasn't it?

Yes, and we could assume that when the client send a request in block1 on a resource on which it already has a pending request, it want deliberately cancel the previous one.

This is exactly like when you put a new value in a map for an existing key. There is 2 implementation choices. You can refuse the addition or replace the old value by the new one.  In the second case, the map remove the old value for you but you can not say that you didn't do this deliberately.

> we also act on behalf of the server in this case by simply canceling the first client's request.

I don't think so. We just said before that cancelling request is be done at client initiative.

>IMHO it would be more informative for the client to receive an (error) response containing a code an payload so that it can better determine the reason why its request failed and react in a more appropriate way.

I'm still thinking this is not a good idea to make "as if" the server send a response.